staging/src/k8s.io/dynamic-resource-allocation/structured/namedresources/cel/compile.go

```diff
@@ -0,0 +1,228 @@
+	}
+	return CompilationResult{
+		Program:     prog,
+		Expression:  expression,
+		OutputType:  ast.OutputType(),
+		Environment: env,
+	}
+}
+
+var valueTypes = map[string]struct {
+	celType *cel.Type
+	// get returns nil if the attribute doesn't have the type, otherwise
+	// the value of that type.
+	get func(attr resourceapi.NamedResourcesAttribute) (any, error)
+}{
+	"quantity": {apiservercel.QuantityType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.QuantityValue == nil {
+			return nil, nil
+		}
+		return apiservercel.Quantity{Quantity: attr.QuantityValue}, nil
+	}},
+	"bool": {cel.BoolType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.BoolValue == nil {
+			return nil, nil
+		}
+		return *attr.BoolValue, nil
+	}},
+	"int": {cel.IntType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.IntValue == nil {
+			return nil, nil
+		}
+		return *attr.IntValue, nil
+	}},
+	"intslice": {types.NewListType(cel.IntType), func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.IntSliceValue == nil {
+			return nil, nil
+		}
+		return attr.IntSliceValue.Ints, nil
+	}},
+	"string": {cel.StringType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.StringValue == nil {
+			return nil, nil
+		}
+		return *attr.StringValue, nil
+	}},
+	"stringslice": {types.NewListType(cel.StringType), func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.StringSliceValue == nil {
+			return nil, nil
+		}
+		return attr.StringSliceValue.Strings, nil
+	}},
+	"version": {SemVerType, func(attr resourceapi.NamedResourcesAttribute) (any, error) {
+		if attr.VersionValue == nil {
+			return nil, nil
+		}
+		v, err := semver.Parse(*attr.VersionValue)
+		if err != nil {
+			return nil, fmt.Errorf("parse semantic version: %v", err)
+		}
+
+		return SemVer{Version: v}, nil
+	}},
+}
+
+var boolType = reflect.TypeOf(true)
+
+func (c CompilationResult) Evaluate(ctx context.Context, attributes []resourceapi.NamedResourcesAttribute) (bool, error) {
+	variables := make(map[string]any, len(valueTypes))
+	for name, valueType := range valueTypes {
+		m, err := buildValueMapper(c.Environment.CELTypeAdapter(), attributes, valueType.get)
+		if err != nil {
+			return false, fmt.Errorf("extract attributes with type %s: %v", name, err)
+		}
+		variables[attributesVarPrefix+name] = m
+	}
+	result, _, err := c.Program.ContextEval(ctx, variables)
+	if err != nil {
+		return false, err
+	}
+	resultAny, err := result.ConvertToNative(boolType)
+	if err != nil {
+		return false, fmt.Errorf("CEL result of type %s could not be converted to bool: %w", result.Type().TypeName(), err)
+	}
+	resultBool, ok := resultAny.(bool)
+	if !ok {
+		return false, fmt.Errorf("CEL native result value should have been a bool, got instead: %T", resultAny)
+	}
+	return resultBool, nil
+}
+
+func mustBuildEnv() *environment.EnvSet {
```

Comment:
cc @jpbetz for eyes on this ... not sure how much of this is copy / paste from other env builders and how much is tweaked to try to allow use of the semver lib in the same version it's introduced

Reply:
The environment should be properly versioned, it's just the env selection which is tweaked. See https://github.com/kubernetes/kubernetes/pull/123516/files#r1516514152

Reply:
This looks like a correct environment augmentation in terms of code structure. I'll comment about the semver library seprately.

Reply:
Once https://github.com/kubernetes/kubernetes/pull/123664 is merge, a full copy of the library is no longer need here. This environment could reference the library implementation defined in core and enable it early in the environment.